### ISSUE TITLE
kconfig/keyval: clarif location(s) of the .load("./.config") file: source_root() or build_root()?

### ISSUE THREAD

**Issue by marc-h38** [mesonbuild/meson#6694]
This issue is basically a "post-merge" review of the initial https://mesonbuild.com/Kconfig-module.html documentation merged in https://github.com/mesonbuild/meson/pull/5031/files. 

It is all about the location of the `keyval.load(".config")` file and only about that. It is focused on the kconfig use case = the only known "keyval" use case for now. I suspect a lot applies outside kconfig but I haven't thought about how much.

I started this discussion in (closed) #5031 because I naively thought it would wrap up quickly. I'll try to make this description relatively self contained but please don't underestimate the devils in the details like I did and make a reasonable effort to read references before commenting. Don't miss the "Show resolved" button in #5031 like I did.

Quoting @bonzini from https://github.com/mesonbuild/meson/pull/5031#issuecomment-590982203
> At this point it's a matter of figuring out the best practices and documenting them. Fortunately none of this is affecting the code of the module itself.

Yes, the main purpose of this issue is to "fix" https://mesonbuild.com/Kconfig-module.html and I _hope_ it won't affect the code but see last section "Bug?" below.

Background: is .config a build input or build output?
-----------------------------------------------------

It depends.

From https://github.com/mesonbuild/meson/pull/5031#issuecomment-590960533
> The main question here is this: is .config a build input or a build output? You can see this pervasive question/confusion all the way from @jpakkane's very first comment in #1617

Now "input" and "output" are relative, for instance meson's output is ninja's input. In this particular issue and context, "build output" is anything the human user is happy deleting because it can/should be regenerated by merely pushing some button(s) without thinking about it. Like this:

https://mesonbuild.com/Using-multiple-build-directories.html
> Secondly this makes it very easy to clean your projects: just delete the build subdirectory and you are done.

Unfortunately, `.config` files can be either build input or build output depending on the use case. Quoting myself from https://github.com/mesonbuild/meson/pull/5031#issuecomment-590960533
> There is confusion because the answer is: it depends. .config can be seen as either. An interactive user who is interactively (and often: painfully) menuconfiguring a single .config at a time would consider it as a build input that she doesn't want to ever be --wiped. For someone or something testing multiple configs, or even someone using a build system generating a single .config from multiple fragments (defconfig, Yocto, chromiumOS, etc.), then .config is indeed a build output that should be --wiped.
> 
> I wouldn't call a "bug" the fact that kconfiglib.py is flexible and convenient enough to automatically support both approaches. But maybe it is ahead of its time and I got carried away by it.

It's not rare to see .config files checkedin [Source Configuration Management](https://en.wikipedia.org/wiki/Software_configuration_management) like git. It's admittedly much more common to see .config _fragments_ there, used to... generate the actual .config!

For even more confusion, .config is both an input and an output... to itself!
https://github.com/zephyrproject-rtos/zephyr/pull/22022/files

Issues with the current documentation
---------------------------------------

Quoting @bonzini's answer at https://github.com/mesonbuild/meson/pull/5031#issuecomment-590982203
> .config is a build input, but it's a build input that can differ for different build roots, hence the original suggestion of placing it (by default) in the build root

OK why not, however this is quite unusual and IMHO uncomfortable for interactive users. So this requires at least better documentation than we currently have. Quoting https://mesonbuild.com/Kconfig-module.html :

> Kconfig frontends usually have ".config" as the default name for the configuration file. However, placing the configuration file in the source directory limits the user to one configuration per source directory.

No, it does not: you can have .config1, .config2, etc. in the same source directory. In fact the first sentence alludes to that and practically contradicts the second sentence. Verified with meson 0.53ish and kconfiglib.py.

More from https://mesonbuild.com/Kconfig-module.html
> In order to allow separate configurations for each build directory, as is the Meson standard, meson.build should not hardcode ".config" as the argument to kconfig.load(), and should instead make the argument to kconfig.load() a project build option.
> 
> The first (and only) argument is the path to the configuration file to load (usually ".config").

Taken in *isolation* from the previous one, this paragraph really seems to recommend the following approach, which works:
```
meson setup   -Ddot_config=source_tree()/config_blue   blue_build_root/
meson setup   -Ddot_config=source_tree()/config_red     red_build_root/
```

Bug?
----

A. if `meson.build_root()/.config` is according to @bonzini the preferred approach, then why does `kconfig.load(".config")` defaults to `meson.source_root()/.config` ?! Shouldn't `kconfig.load(".config")` default to `meson.build_root()/.config` instead? Also, why do the tests currently in `test cases/kconfig` either load .config directly from source_root() or copy from it?

B. Let's pretend `kconfig.load(".config")` is changed and starts defaulting to `meson.build_root()/.config`. Should it still fallback to `meson.source_root()/.config` so single config interactive users keep the convenience of `--wiping` their build directory? Or should it never look at `meson.source_root()/.config` ever more?

C. If fallback B. is implemented, should it warn when both files are present?

**Comment by bonzini** [comment]
The answer to the question is simple, it uses the source directory for consistency with pretty much every other case where a string in `meson.build` is interpreted as a file.

**Comment by marc-h38** [comment]
Thanks @bonzini , this answers the "why" but it doesn't answer the "Should it?". Consistency between the documentation and implementation of this particular module seems more important to me than consistency with other, unrelated features.

At the very least the documentation should warn about this inconsistency.

**Comment by bonzini** [comment]
>  between the documentation and implementation of this particular module seems more important to me than consistency with other, unrelated features.

What is the inconsistency? The documentation says that

```md
placing the configuration file in the source
directory limits the user to one configuration per source directory.
In order to allow separate configurations for each build directory, as is
the Meson standard, `meson.build` should not hardcode ".config" as the
argument to `kconfig.load()`,
```
... because that would load the configuration file from the source directory.

**Comment by marc-h38** [comment]
> ... because that would load the configuration file from the source directory.

Not necessarily. See `src/config_blue` and `src/config_red` example above.

> What is the inconsistency? 

- As I just discovered, the recommendation is apparently `build_root/.config`
- The default value, the tests and half the documentation point to `source_root/`.

**Comment by bonzini** [comment]
> As I just discovered, the recommendation is apparently `build_root/.config`

No, the recommendation is an option. `src/config_blue` and `src/config_red` work because you have an option, and they are still "separate configurations for each build directory" (which the user chose to place somewhere in the source directory, but neither Meson nor `meson.build` really care).

**Comment by gregsskyles** [comment]
Are we just hung up on the phrase 'as is the Meson standard'? It does seem to imply some level of recommendation or standardization. Perhaps this section of the documentation can state that there is an option or choice, something like:

The user has the option to have one `.config` file for the source directory, or one `.config` file per build directory (the typical Meson setup), depending on their use case. If you intend to generate and manage multiple configurations from the same source tree, the latter option fits; otherwise, the former will suit and is simpler.

The first (and only) argument to kconfig.load() is the path to the configuration file to load (usually just ".config"). This will result in a single `.config` file in the root of the source tree.

In order to allow separate configurations for each build directory, meson.build can make the argument to kconfig.load() a project build option.

**Comment by marc-h38** [comment]
Thanks @gregsskyles for the feedback, appreciated.

I've done more testing and thinking and I believe the documentation should take a stronger stance towards `meson.source_root()` and discourage `meson.build_root() / '.config'`.

In other words, `.config` should always be considered as a build **input** by meson.

For instance this sentence in the documentation should be changed or removed:

> However, placing the configuration file in the source directory limits the user to one configuration per source directory.

It should be removed first because it's incorrect (`src/config_blue` and `src/config_red` work perfectly) and it should also be removed because `meson.build_root() / '.config'` either doesn't make sense, or is very impractical, or both.

- Either `.config` is _manually menuconfigured by the user and it makes no sense to have it in a directory that can be `--wiped`.

- `meson.build_root()` is more appealing for the case where `.config` is generated by some tool (from config fragments or whatever). However I tried this and it simply doesn't work because it's impossible to generate it with a `custom_target()` BEFORE `kconfig.load()`ing it.

**Comment by bonzini** [comment]
> For instance this sentence in the documentation should be changed or removed:
>
> > However, placing the configuration file in the source directory limits the user to one configuration per source directory.

I agree with this, for example: "However, hard coding the configuration file to a fixed name in the source directory limits the user to one configuration per source directory."

> However I tried this and it simply doesn't work because it's impossible to generate it with a custom_target() BEFORE kconfig.load()ing it

You can do that with `configure_file`. QEMU does it.

**Comment by marc-h38** [comment]
> You can do that with configure_file. QEMU does it.

**Comment by marc-h38** [comment]
> You can do that with configure_file. QEMU does it.

Thanks, I missed `configure_file` has a `command` feature! This does make a big difference.

So here's my new recommendation for the documentation:
- Recommend the `source_root` directory location (the default) for  manually menuconfigured .config/config_blue files. For either single or multiple build directories. So they never get accidentally deleted.
- Recommend explicit `build_root()/.config` directory + `configure_file( command` for  .config files machine generated from config fragments/defconfigs/etc. For either single or multiple build directories.

If everyone agrees I'll try to find the time to submit a documentation update (but please beat me to it :-)
